Add documentation for classes. (#1580)
pyccel/pyccel: docs/classes.md

# Supported Class by Pyccel

> Pyccel strives to provide robust support for object-oriented programming concepts commonly used by developers. In Pyccel, classes are a fundamental building block for creating structured and reusable code. This documentation outlines key features and considerations when working with classes in Pyccel.

Reviewer: ```suggestion
Pyccel strives to provide robust support for object-oriented programming concepts commonly used by developers. In Pyccel, classes are a fundamental building block for creating structured and reusable code. This documentation outlines key features and considerations when working with classes in Pyccel.

:warning: Pyccel's class support is currently limited to translations only. The implementation of the wrapper which will make the class accessible from Python is in progress.
```

## Constructor Method

-   The Constructor Method, `__init__`, is used to initialize the object's attributes.

> -   It must be named `__init__`, or the Pyccel compiler will generate an error.

Reviewer: What error are you talking about here? Normally it is not possible to initialise an object's attributes with a function with a different name. Is your point that you have to have an `__init__` method or that you cannot assign to attributes in other methods if they have not first been created in the `__init__` method?

Author: Fixed https://github.com/pyccel/pyccel/pull/1580/commits/f92c3e1883b231711e3cbad1f475439610d22f96.

-   The first parameter of the `Constructor Method` should always be named `self`.

> -   Pyccel doesn't support Interface Constructor.

Reviewer: The term `Interface` comes from Fortran. I'm not sure it is widely understood by our users. I suggest:
```suggestion
-   Pyccel only permits one type definition for each of the arguments of the `__init__` method. Union types or templates cannot be used.
```

## Destructor Method

-   The Destructor Method, `__del__`, is used to perform cleanup actions when an object is destroyed.

> -   Pyccel automatically adds attributes that need to be freed to the garbage collector. These attributes are subsequently released during the destructor's execution. This mechanism helps manage memory efficiently, ensuring that resources are appropriately deallocated.
> -   One notable feature of Pyccel is its automatic management of destructor calls, even when constructors are invoked in different scopes. This ensures that resources are properly cleaned up when class instances are no longer in use.

Reviewer: I'm not sure this is notable. It is critical to avoid seg faults though.
```suggestion
-   Pyccel automatically takes care of garbage collection for classes.
    - Attributes are released during the destructor's execution.
    - The class destructor is called automatically once the class goes out of scope.
```

> -   The first parameter of the `Destructor Method` should always be named `self`.

Reviewer: Isn't this true for any class method not just for `__init__` and `__del__`?

Reviewer: I've just seen this is mentioned below. I don't know if it's worth repeating 3 times. Especially as you are already working on removing this restriction (#1479) so you will need to delete this in 3 places. If you don't mind deleting it 3 times though then why not.

Author: Fixed https://github.com/pyccel/pyccel/pull/1580/commits/f92c3e1883b231711e3cbad1f475439610d22f96.

```

### - C Program File Equivalent
```C
int main()
{
    t_ndarray Dummy_0000 = {.shape = NULL};
    struct MyClass obj;
    Dummy_0000 = array_create(1, (int64_t[]){INT64_C(1)}, nd_double, false, order_c);
    array_fill((double)1.0, Dummy_0000);
    MyClass__init__(&obj, INT64_C(1), Dummy_0000);
    free_array(&Dummy_0000);
    MyClass__del__(&obj);
    return 0;
}
```

### C/VALGRIND _OUTPUT_

```Shell
Object created!
1 [1.000000000000]

==151606== Memcheck, a memory error detector
==151606== Copyright (C) 2002-2017, and GNU GPL'd, by Julian Seward et al.
==151606== Using Valgrind-3.18.1 and LibVEX; rerun with -h for copyright info
==151606== Command: ./MyClass
==151606==
Object created!
1 [1.000000000000]
==151606==
==151606== HEAP SUMMARY:
==151606==     in use at exit: 0 bytes in 0 blocks
==151606==   total heap usage: 6 allocs, 6 frees, 1,064 bytes allocated
==151606==
==151606== All heap blocks were freed -- no leaks are possible
==151606==
==151606== For lists of detected and suppressed errors, rerun with: -s
==151606== ERROR SUMMARY: 0 errors from 0 contexts (suppressed: 0 from 0
```

## Class Methods

-   Pyccel now supports Class Methods and Interfaces.

> -   In Pyccel, class attributes can be initialized within any method of the class.

Reviewer: Is this true? Doesn't it contradict what you say about the constructor?

Author: Yeah normally `__init__` where we have to initilise them but we can create other attributes of a class in any other method of the same class.

-   Pyccel enables classes to be passed as arguments to methods and functions and returned with modified data.

> -   The first parameter of the Class Method should always be named `self`.

Reviewer: I think it may be worth moving this further up the file and grouping the examples. As they are quite detailed it could be easy to miss this section in the middle

Author: Fixed https://github.com/pyccel/pyccel/pull/1580/commits/65baee9d82decc0a4f2026777e35685a2ded628a.
Fixed https://github.com/pyccel/pyccel/pull/1580/commits/2cbad56fd8f824b2e827ac6512a991e661baa2a9.

}
/*........................................*/
/*........................................*/
void MyClass1__Method1(struct MyClass1* self, struct MyClass* param1)
{
    self->param = (*param1);
}
/*........................................*/
/*........................................*/
struct MyClass MyClass1__Method2(struct MyClass1* self)
{
    struct MyClass Out_0001;
    MyClass__init__(&Out_0001, INT64_C(1));
    return Out_0001;
}
/*........................................*/
/*........................................*/
void MyClass1__del__(struct MyClass1* self)
{
    if (!self->is_freed)
    {
        // pass
        MyClass__del__(&self->param);
        self->is_freed = 1;
    }
}
/*........................................*/
```

### - C Program File Equivalent
```C
int main()
{
    struct MyClass1 obj;
    struct MyClass Dummy_0000;
    MyClass1__init__(&obj);
    Dummy_0000 = MyClass1__Method2(&obj);
    MyClass1__Method1(&obj, &Dummy_0000);
    printf("%"PRId64"\n", obj.param.param1);
    MyClass1__del__(&obj);
    return 0;
}
```

### - C _OUTPUT_

```Shell
MyClass1 Object created!
MyClass Object created!
1
```

## Limitations

> It's important to note that Pyccel does not support class wrappers, inheritance, properties, class variables, or inline class methods. The focus of Pyccel is primarily on core class functionality and memory management.

Reviewer: Have you tested inline class methods? I plan to use this soon if it doesn't work then I'll need to fix it first :sweat_smile:

Author: Fixed https://github.com/pyccel/pyccel/pull/1580/commits/3e0def115492fb89aae4d30826365da820416f87.
I need also to add tests in https://github.com/pyccel/pyccel/pull/1576.

> It's important to note that Pyccel does not support class wrappers, inheritance, properties, class variables, or inline class methods. The focus of Pyccel is primarily on core class functionality and memory management.

Reviewer: I think we should mention the wrappers at the top, especially while we're developing the wrapper. We can then delete the line once we're done.

Author: Fixed https://github.com/pyccel/pyccel/pull/1580/commits/4f63dd4707a49ee3652c0eb212dad688eed1e600.

> It's important to note that Pyccel does not support class wrappers, inheritance, properties, class variables, or inline class methods. The focus of Pyccel is primarily on core class functionality and memory management.

Reviewer: ```suggestion
It's important to note that Pyccel does not support class wrappers, inheritance, properties, magic methods, class variables, or inline class methods. The focus of Pyccel is primarily on core class functionality and memory management.
```

> It's important to note that Pyccel does not support class wrappers, inheritance, properties, class variables, or inline class methods. The focus of Pyccel is primarily on core class functionality and memory management.

Reviewer: ```suggestion
It's important to note that Pyccel does not support class wrappers, inheritance, properties, class variables, or inline class methods. For our first implementation, the focus of Pyccel is primarily on core class functionality and memory management.
```

Author: Fixed https://github.com/pyccel/pyccel/pull/1580/commits/2467c0d7adf8ac3547ccf5ce4edc7ad2c6891083.
